src/data-source/collection-browser-data-source.ts

```diff
@@ -0,0 +1,1162 @@
+    // If so, we just want to discard this set of aggregations because they are
+    // likely no longer valid for the newer query.
+    const queryChangedSinceFetch =
+      facetFetchQueryKey !== this.facetFetchQueryKey;
+    if (queryChangedSinceFetch) return;
+
+    if (!success) {
+      const errorMsg = searchResponse?.error?.message;
+      const detailMsg = searchResponse?.error?.details?.message;
+
+      if (!errorMsg && !detailMsg) {
+        // @ts-ignore: Property 'Sentry' does not exist on type 'Window & typeof globalThis'
+        window?.Sentry?.captureMessage?.(
+          'Missing or malformed facet response from backend',
+          'error'
+        );
+      }
+
+      return;
+    }
+
+    const { aggregations, collectionTitles } = success.response;
+    this.aggregations = aggregations;
+
+    if (collectionTitles) {
+      for (const [id, title] of Object.entries(collectionTitles)) {
+        this.collectionTitles.set(id, title);
+      }
+    }
+
+    this.yearHistogramAggregation =
+      success?.response?.aggregations?.year_histogram;
+
+    this.host.setFacetsLoading(false);
+    this.host.requestUpdate();
+  }
+
+  /**
+   * Performs the initial page fetch(es) for the current search state.
+   */
+  private async doInitialPageFetch(): Promise<void> {
+    this.host.setSearchResultsLoading(true);
+    // Try to batch 2 initial page requests when possible
+    await this.fetchPage(this.host.initialPageNumber, 2);
+    this.host.setSearchResultsLoading(false);
+  }
+
+  /**
+   * Fetches one or more pages of results and updates the data source.
+   *
+   * @param pageNumber The page number to fetch
+   * @param numInitialPages If this is an initial page fetch (`pageNumber = 1`),
+   *  specifies how many pages to batch together in one request. Ignored
+   *  if `pageNumber != 1`, defaulting to a single page.
+   */
+  async fetchPage(pageNumber: number, numInitialPages = 1): Promise<void> {
+    console.log(
+      `fetchPage(${pageNumber})`,
+      this.canPerformSearch,
+      this.hasPage(pageNumber),
+      this.endOfDataReached,
+      this.host.baseQuery,
+      JSON.stringify(this.host.selectedFacets)
+    );
+    const trimmedQuery = this.host.baseQuery?.trim();
+    if (!this.canPerformSearch) return;
+
+    // if we already have data, don't fetch again
+    if (this.hasPage(pageNumber)) return;
+
+    if (this.endOfDataReached) return;
+
+    // Batch multiple initial page requests together if needed (e.g., can request
+    // pages 1 and 2 together in a single request).
+    const numPages = pageNumber === 1 ? numInitialPages : 1;
+    const numRows = this.pageSize * numPages;
+
+    // if a fetch is already in progress for this query and page, don't fetch again
+    const { pageFetchQueryKey } = this;
+    const pageFetches =
+      this.pageFetchesInProgress[pageFetchQueryKey] ?? new Set();
+    if (pageFetches.has(pageNumber)) {
+      console.log(
```

Comment:
we can introduce `log()` inside collection-browser as well for logs on dev instances.

Reply:
Thanks, I've ported in the `log()` util function from Offshoot. Most of the log statements are removed now, but a few that may be helpful for further debugging now use this function instead of bare `console.log`.